### PR TITLE
Use ctime of dismissal as dtime for items

### PR DESCRIPTION
**Pull request by patrickxb**
r? @maxtaco 

I haven't come up with a clean way to test this change.  The existing tests pass, but if you have any ideas, please let me know.

### CODE REVIEW

File: protocol/avdl/common.avdl [hunk 1]
```diff
@@ -38,6 +38,7 @@ protocol common {
 	record Dismissal {
 		array<MsgID> msgIDs;
 		array<MsgRange> ranges;
+		Time ctime;
```
Comment (maxtaco) [inline protocol/avdl/common.avdl]: This shouldn't be necessary. The dismissal always comes with a Metadata, which should have a CTime in it.

File: storage/schema.go [hunk 2]
```diff
@@ -42,6 +42,7 @@ var schema = []string{
 		uid   CHAR(16) NOT NULL,
 		msgid CHAR(16) NOT NULL,
 		dmsgid CHAR(16) NOT NULL, -- "the message IDs to dismiss",
+		ctime DATETIME(6) NOT NULL,
```
Comment (maxtaco) [inline storage/schema.go]: this also shouldn't be necessary (following the above comment)

**Comment by patrickxb** [comment]
@maxtaco PTAL

File: storage/schema.go [hunk 3]
```diff
@@ -51,6 +51,7 @@ var schema = []string{
 		msgid CHAR(16) NOT NULL,
 		category VARCHAR(128) NOT NULL,
 		dtime DATETIME(6) NOT NULL, -- "throw out matching events before dtime",
+		ctime DATETIME(6) NOT NULL,
```
Comment (maxtaco) [inline storage/schema.go]: can you nuke this?

File: storage/sql_sm.go [hunk 4]
```diff
@@ -127,29 +133,36 @@ func (s *SQLEngine) consumeCreation(tx *sql.Tx, u gregor.UID, i gregor.Item) err
...
-	qb.Build("WHERE uid=? AND msgid=?")
-	upd, err := tx.Prepare(qb.Query())
+	upd, err := tx.Prepare("UPDATE items SET dtime=? WHERE uid=? AND msgid=?")
 	if err != nil {
 		return err
 	}
 	defer upd.Close()
+
+	var ct time.Time
+	if ctime != nil && ctime.Time() != nil {
```
Comment (maxtaco) [inline storage/sql_sm.go]: does this case ever happen?

File: storage/sql_sm.go [hunk 5]
```diff
@@ -127,29 +133,36 @@ func (s *SQLEngine) consumeCreation(tx *sql.Tx, u gregor.UID, i gregor.Item) err
...
+		ct = *ctime.Time()
+	} else {
+		ct = s.clock.Now()
```
Comment (maxtaco) [inline storage/sql_sm.go]: I've intentionally avoided this so that we can use the MySQL's server clock.

File: storage/sql_sm.go (hunk 4 above)
Comment (maxtaco) [inline storage/sql_sm.go]: I would expected `ctime` to always be non-nil here.

File: storage/sql_sm.go [hunk 6]
```diff
@@ -226,32 +258,37 @@ func (s *SQLEngine) consumeInBandMessage(m gregor.InBandMessage) error {
...
+		if err != nil {
+			tx.Rollback()
+		} else {
+			err = tx.Commit()
+		}
+	}()
+
 	md := m.Metadata()
-	if err := s.consumeInBandMessageMetadata(tx, md, gregor.InBandMsgTypeUpdate); err != nil {
+	if err = s.consumeInBandMessageMetadata(tx, md, gregor.InBandMsgTypeUpdate); err != nil {
```
Comment (maxtaco) [inline storage/sql_sm.go]: I thought you'd be setting ctime here if it wasn't already set.

Comment (maxtaco) [inline storage/sql_sm.go]: (Inside `consumeInBandMessageMetadata`)

File: storage/sql_sm.go (hunk 5 above)
Comment (patrickxb) [inline storage/sql_sm.go]: I was just putting something in if it is empty, which seems to happen in the tests.

File: storage/sql_sm.go [hunk 7]
```diff
@@ -157,21 +170,40 @@ func (s *SQLEngine) consumeMsgIDsToDismiss(tx *sql.Tx, u gregor.UID, mid gregor.
...
 		qb.TimeOrOffset(mr.EndTime())
+		qb.Build(",")
+		qb.Now()
 		qb.Build(")")
 		if err := qb.Exec(tx); err != nil {
 			return err
 		}
 
+		// get ctime from the dismissal row
+		ctime, err := s.ctimeFromDismissalsByTime(tx, u, mid, mr)
```
Comment (maxtaco) [inline storage/sql_sm.go]: I'm also not certain why this is needed. Can't `ctime` be passed in as in the corresponding calls to `consumeMsgIDsToDismiss`?

File: storage/sql_sm.go (hunk 5 above)
Comment (patrickxb) [inline storage/sql_sm.go]: (now that it is getting it from metadata)

Comment (maxtaco) [inline storage/sql_sm.go]: You should ensure that `ctime` is always set in the metadata first (see comment in this PR somewhere)

File: storage/schema.go (hunk 3 above)
Comment (patrickxb) [inline storage/schema.go]: no, it's needed to let mysql set the time on the insert.

Comment (maxtaco) [inline storage/schema.go]: It shouldn't be, `ctime` should be set in the metadata row, not this row.

File: storage/sql_sm.go (hunk 7 above)
Comment (patrickxb) [inline storage/sql_sm.go]: yes.

from our slack discussion yesterday, I was under the impression that the insert into dismissals_* needed to create the ctime, then select from it.  obviously I missed something.

Comment (maxtaco) [inline storage/sql_sm.go]: right, but i was trying to say that all of this was in the `consumeMetadata` function.

**Comment by patrickxb** [comment]
Ok, let's try again...PTAL

File: storage/sql_sm.go [hunk 8]
```diff
@@ -127,29 +138,29 @@ func (s *SQLEngine) consumeCreation(tx *sql.Tx, u gregor.UID, i gregor.Item) err
...
-	qb.Build("WHERE uid=? AND msgid=?")
-	upd, err := tx.Prepare(qb.Query())
+	upd, err := tx.Prepare("UPDATE items SET dtime=? WHERE uid=? AND msgid=?")
 	if err != nil {
 		return err
 	}
 	defer upd.Close()
+
+	qb := s.newQueryBuilder()
+	ctimeArg := qb.TimeArg(ctime)
```
Comment (maxtaco) [inline storage/sql_sm.go]: collapse into one line?

File: storage/sql_sm.go [hunk 9]
```diff
@@ -127,29 +138,29 @@ func (s *SQLEngine) consumeCreation(tx *sql.Tx, u gregor.UID, i gregor.Item) err
...
+	hexUID := hexEnc(u)
+	hexMID := hexEnc(mid)
+
 	for _, dmid := range dmids {
-		_, err = ins.Exec(hexEnc(u), hexEnc(mid), hexEnc(dmid))
+		_, err = ins.Exec(hexUID, hexMID, hexEnc(dmid))
 		if err != nil {
 			return err
 		}
-		qbc := qb.Clone()
```
Comment (maxtaco) [inline storage/sql_sm.go]: can get rid of this function now

File: storage/sql_sm.go [hunk 10]
```diff
@@ -203,9 +220,28 @@ func (s *SQLEngine) consumeInBandMessageMetadata(tx *sql.Tx, md gregor.Metadata,
 	qb := s.newQueryBuilder()
 	qb.Build("INSERT INTO messages(uid, msgid, mtype, devid, ctime) VALUES(?, ?, ?, ?,",
 		hexEnc(md.UID()), hexEnc(md.MsgID()), int(t), hexEncOrNull(md.DeviceID()))
-	qb.Now()
+	if md.CTime().IsZero() {
```
Comment (maxtaco) [inline storage/sql_sm.go]: will this ever be 0?

Comment (maxtaco) [inline storage/sql_sm.go]: or course it will be, i missed the collapse point above. sorry!

File: test/objects.go [hunk 11]
```diff
@@ -219,6 +221,10 @@ func (t *testDismissal) RangesToDismiss() []gregor.MsgRange {
 	return ret
 }
 
+func (t *testDismissal) CTime() time.Time {
```
Comment (maxtaco) [inline test/objects.go]: can we nuke this?

**Comment by maxtaco** [comment]
:+1:

**Pull request merged**